ykrt/src/compile/jitc_yk/jit_ir.rs

```diff
@@ -333,6 +428,118 @@ impl Module {
     pub(crate) fn dump(&self) {
         eprintln!("{}", self);
     }
+
+    /// Push a slice of extra arguments into the extra arg table.
+    fn push_extra_args(&mut self, ops: &[Operand]) -> ExtraArgsIndex {
+        let idx = self.extra_args.len();
+        self.extra_args.extend_from_slice(ops); // FIXME: this clones.
+        ExtraArgsIndex(u16::try_from(idx).unwrap()) // FIXME: propagate error
+    }
+
+    /// Lookup a type by its index. Panics if an out of bounds index is supplied.
+    fn get_type(&self, tidx: TypeIndex) -> &Type {
+        &self.types[usize::from(tidx.0)]
+    }
+
+    /// Push a new type into the type table and return its index.
+    fn push_type(&mut self, ty: Type) -> TypeIndex {
+        let idx = self.types.len();
+        self.types.push(ty);
+        TypeIndex(u16::try_from(idx).unwrap()) // FIXME: propagate error
+    }
+
+    /// Push a new constant into the constant table and return its index.
+    pub(crate) fn push_const(&mut self, constant: Constant) -> ConstIndex {
+        let idx = self.consts.len();
+        self.consts.push(constant);
+        ConstIndex(u16::try_from(idx).unwrap()) // FIXME: propagate error
+    }
+
+    /// Get the index of a type, inserting it in the type table if necessary.
+    pub fn const_index(&mut self, c: &Constant) -> ConstIndex {
+        // FIXME: can we optimise this?
```

Comment:
How often might this function be called? If "a lot" then "use a hash table" might end up being the right advice. But for now a linear search is probably OK.

Reply:
Every time you "intern" a type. We can't expect the API user to know if a type has already been interned, so we do the check for them.

From experience with LLVM, you throw types at a module quite frequently and willy-nilly, so perhaps we should in fact have some kind of cache, but should we do that as another PR?

Reply:
> I think we opted to not share constants with the AOT IR for now?

We agreed the opposite, but I'm not completely wedded to that (ideally we don't want to do things like copy large constant strings because it's a waste) provided you're reasonably sure that changing this later wouldn't be difficult.

Reply:
What luajit does, it has a constant array, and then stores only a size and len into it. We could do this?

Reply:
The challenge with "constants" is that they're not all present in source code. If you write `x = 2 + 3` you'll probably generate at compile-time the constant `5` even though it's not in the source.

Reply:
> I'm more interested in the kinds of constant folding that can only be done by the JIT due to speculation.

That's my point!

In that sense, AOT constants and JIT constants *could* (but don't have to be!) different things. It's not something we *have* to get right now, but we should remain open to both possibilities IMHO.

Reply:
OK, well for now, the JIT has its own constant table and doesn't share any constants the AOT module. That probably wouldn't be hard to change later.

What I think will bother you is that, due to the above, we still have the "find it or insert it" idiom (that you didn't like) for constants. We were able to kill that idiom for functions by sharing with AOT, but that's not what we are doing here.

Another way around this is to duplicate constants at the cost of memory. I don't like it though. I imagine some constants will come up time and time again like `0` and `1` and `-1`. If we duplicate constants, we also can't compare them for equality by a simple index comparison.

Reply:
OK let's move on.